jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -336,63 +336,75 @@ private Runnable reservedWait()
         @Override
         public void run()
         {
-            while (isRunning())
+            try
             {
-                // test and increment size BEFORE decrementing pending,
-                // so that we don't have a race starting new pending.
-                int size = _size.get();
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} Entering", this);
 
-                // Are we stopped?
-                if (size < 0)
-                    return;
-
-                // Are we surplus to capacity?
-                if (size >= _capacity)
+                while (isRunning())
                 {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("{} size {} > capacity", this, size, _capacity);
-                    if (_starting)
-                        _pending.decrementAndGet();
-                    return;
-                }
+                    // test and increment size BEFORE decrementing pending,
+                    // so that we don't have a race starting new pending.
+                    int size = _size.get();
```

Comment:
my preferred change here is to make size and pending an `AtomicBiInteger`.  I have already done that work in a branch against jetty-10, but have not put it forward as I wanted to see if we can work out the other issues first.    But if we are going to make changes, then I'd prefer to make the logic clearly atomic with a BiInteger rather than optimisticly atomic.

Reply:
We had a lengthy discussion on the subject yesterday with @sbordet. Our conclusion was that the `pending` counter seems useless if not for reporting, and the `size` counter could be replaced with `_stack.size()`.

There's always going to be some non-atomicity between the counters and the stack, so I'm not sure what making both counters atomic between themselves would bring.

Reply:
You can't do that.  It used to be like that at it was a race.    See my PR as I think it can be done without any race.